src/backend/impl/src/services/proposal_service.rs

```diff
@@ -0,0 +1,284 @@
+};
+
+const NNS_GOVERNANCE_CANISTER_ID: &str = "rrkah-fqaaa-aaaaa-aaaaq-cai";
+
+#[cfg_attr(test, mockall::automock)]
+pub trait ProposalService {
+    fn get_proposal(&self, id: ProposalId) -> Result<GetProposalResponse, ApiError>;
+
+    fn get_proposals(&self) -> Result<GetProposalsResponse, ApiError>;
+
+    fn update_proposal_state(
+        &self,
+        id: ProposalId,
+        state: ReviewPeriodState,
+    ) -> Result<(), ApiError>;
+
+    async fn fetch_and_save_nns_proposals(&self) -> Result<(), ApiError>;
+}
+
+pub struct ProposalServiceImpl<T: ProposalRepository> {
+    proposal_repository: T,
+}
+
+impl Default for ProposalServiceImpl<ProposalRepositoryImpl> {
+    fn default() -> Self {
+        Self::new(ProposalRepositoryImpl::default())
+    }
+}
+
+impl<T: ProposalRepository> ProposalService for ProposalServiceImpl<T> {
+    fn get_proposal(&self, id: ProposalId) -> Result<GetProposalResponse, ApiError> {
+        let proposal = self
+            .proposal_repository
+            .get_proposal_by_id(&id)
+            .ok_or_else(|| {
+                ApiError::not_found(&format!("Proposal with id {} not found", &id.to_string()))
+            })?;
+
+        Ok(map_get_proposal_response(id, proposal))
+    }
+
+    fn get_proposals(&self) -> Result<GetProposalsResponse, ApiError> {
+        let proposals = self
+            .proposal_repository
+            .get_proposals()
+            .into_iter()
+            .map(|(id, proposal)| map_get_proposal_response(id, proposal))
+            .collect();
+
+        Ok(GetProposalsResponse { proposals })
+    }
+
+    fn update_proposal_state(
+        &self,
+        id: ProposalId,
+        state: ReviewPeriodState,
+    ) -> Result<(), ApiError> {
+        let mut proposal = self
+            .proposal_repository
+            .get_proposal_by_id(&id)
+            .ok_or_else(|| {
+                ApiError::not_found(&format!("Proposal with id {} not found", &id.to_string()))
+            })?;
+
+        match (proposal.state, state.clone()) {
+            (ReviewPeriodState::InProgress, ReviewPeriodState::Completed) => {
+                proposal.state = state;
+
+                self.proposal_repository.update_proposal(&id, proposal)
+            }
+            _ => Err(ApiError::invalid_argument(
+                "Invalid proposal state transition",
+            )),
+        }
+    }
+
+    async fn fetch_and_save_nns_proposals(&self) -> Result<(), ApiError> {
+        let nns_governance_canister =
+            GovernanceCanisterService(Principal::from_str(NNS_GOVERNANCE_CANISTER_ID).unwrap());
+
+        let ListProposalInfoResponse {
+            proposal_info: proposals,
+        } = match nns_governance_canister
+            .list_proposals(ListProposalInfo {
+                include_reward_status: vec![],
+                omit_large_fields: Some(true),
+                before_proposal: None,
+                limit: 20,
```

Comment:
If there's more than 20 proposals open at the same time, we would miss some. I don't know usually how many System Canister Management proposals are open at once, but I'd be more worried about that than Replica Version Management. We could increase this to a really high number, or do a more complex approach with pagination. Either way we can keep it as-is for now and improve it later, but something to think about in the meantime I think.

Reply:
I think we won't incur in this problem because the proposals polling is done every 5 minutes. I'll anyway increase it to 50 ([the limit is 100](https://github.com/dfinity/ic/blob/293b694571b5bd17b0858c978543b778159f75eb/rs/nns/governance/src/gen/ic_nns_governance.pb.v1.rs#L2663-L2665)).